File: extsort.js
#!/usr/bin/env node

/*
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

var fs = require('fs-extra'); //npm install fs-extra
const path = require('path');

function PrintHelp() {
	console.log("Usage: " + __filename + " [options] <dir(s)> <file(s)>\n"
		+ "	-R     | Be recursive.\n"
		+ "	--help | Print this help.\n\n"
		+ "Example: " + __filename + " /user/junk\n");
	process.exit(-1);
}

if(process.argv.length <= 2) {
	PrintHelp();
}

var recursive = false;

function ProcessFile(full_file) {
	//Store basename
	var BaseFileName = path.basename(full_file);

	//Make sure we aren't working with a dotfile.
	if(BaseFileName.charAt(0) != '.') {
		var ext = path.extname(full_file).slice(1);
		var CurrentDir = path.dirname(full_file);
		var NewDir = path.join(CurrentDir,ext);
		var NewFileLocation = path.join(NewDir,BaseFileName);
		
		fs.ensureDir(NewDir).then(() => {
			fs.move(full_file, NewFileLocation, { overwrite: true }).then(() => {
				console.log('Moved ' + full_file + ' => ' + NewFileLocation);
			}).catch(err => {
				console.error(err);
			});
		}).catch(err => { 
			console.error(err);
		});
	}
}

function ProcessPathArgs(current_path, basedir) {
	var IsRecursiveValid = () => {
		return(basedir || recursive);
	}

	//Function to check the details of the arguemtn provided
	fs.lstat(current_path).then(stats => {
		//Take this logic if the path is determined to be a file.
		if(stats.isFile()) {
			ProcessFile(current_path);
		}
		
		//Check if recursion is valid.
		else if(stats.isDirectory() && IsRecursiveValid()) {
			//Read file list into array.
			fs.readdir(current_path).then(files => {
				//Feed every file into the ProcessFile() function.
				files.forEach(file => {
					ProcessFile(path.join(current_path,file),false);
				});
			}).catch(err => {
				console.error(err);
			});
		} else {
			console.error(current_path + ' is not a file or directory.');
		}
	}).catch(err => {
		console.error(err);
	});
}

function ProcessArgs() {
	var args = new Set(process.argv.slice(2));
	
	//If the help command was given at all, we just do help and quit.
	if(args.has('--help')) {
		PrintHelp();
	}

	//Set variable for dry run.
	if(args.has('-R')) {
		recursive = true;
		args.delete('-R')
	}

	args.forEach(current_arg => {
		ProcessPathArgs(path.resolve(current_arg),true);
	});
}

ProcessArgs();

File: extsort.cpp
/*
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <filesystem>
#include <functional>
#include <cstring>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <libioapp.hpp>

//Globals
bool recursive = false;

std::string RemoveFirstChar(std::string s)
{
	s.erase(s.begin());
	return s;
}

int PrintHelp(void)
{
	std::cerr <<
	"Usage: extsort [options] <dir(s)> <file(s)>\n" <<
	"	-R     | Recursively navigate directories.\n" <<
	"	--help | Print this help.\n\n" <<
	"Example: extsort /user/junk /mnt/stuff/things.txt\n";
	return 1;
}

int main(int argc, char* argv[])
{
	if(argc == 0)
	{
		return PrintHelp();
	}

	ioapp::ProcessArgs extsort_ProcessArgs(argc,argv,true,1);

	for(const std::string &f : extsort_ProcessArgs.Flags)
	{
		if(f == "--help")
		{
			return PrintHelp();
		}

		if(f == "-R")
		{
			recursive = true;
		}
	}

	for(const std::string &arg : extsort_ProcessArgs.Paths)
	{
		ioapp::ioapp(std::filesystem::path(arg), [&](std::filesystem::path p) {
			if(p.has_extension())
			{
				std::string new_subdir = p.parent_path().string() + '/' + RemoveFirstChar(p.extension().string());
				std::string new_file = new_subdir + '/' + p.filename().string();

				//If the subdir we plan to move to exists.
				if(!std::filesystem::is_directory(new_subdir))
				{
					if(std::filesystem::create_directory(new_subdir))
					{
						//We try two methos of moving a file, and crash if they fail.
						try
						{
							std::filesystem::rename(p, new_file);
						}

						catch(const std::exception& e)
						{
							try
							{
								std::filesystem::copy(p, new_file);
								std::filesystem::remove(p);
							}

							catch(const std::exception& ex)
							{
								std::cerr << ex.what() << '\n' << "Failed to move " << p.string() << std::endl;
								exit(1);
							}
						}

						std::cout << "Move: " << p.string() << " -> " << new_file << std::endl;
					}

					else
					{
						std::cerr << "Could not create directory: " << new_subdir << '\n';
						exit(1);
					}
				}
			}
		}, recursive);
	}

	return 0;
}